draft-ietf-quic-transport.md

```diff
@@ -1549,6 +1543,42 @@ disable_migration (0x0009):
   address other than that used to perform the handshake.  This parameter is a
   zero-length value.
 
+Either peer MAY advertise an initial value for the flow control on each type of
+stream on which they might receive data.  Each of the following transport
+parameters is encoded as an unsigned 32-bit integer in units of octets:
+
+initial_max_stream_data_bidi_local (0x0000):
+
+: The initial stream maximum data for bidirectional, locally-initiated streams
+  parameter contains the initial flow control limit for newly created
+  bidirectional streams opened by the endpoint that sets the transport
+  parameter.  In client transport parameters, this applies to streams with an
+  identifier ending in 0x0; in server transport parameters, this applies to
+  streams ending in 0x1.
+
+initial_max_stream_data_bidi_remote (0x000a):
+
+: The initial stream maximum data for bidirectional, peer-initiated streams
+  parameter contains the initial flow control limit for newly created
+  bidirectional streams opened by the endpoint that receives the transport
+  parameter.  In client transport parameters, this applies to streams with an
+  identifier ending in 0x1; in server transport parameters, this applies to
+  streams ending in 0x0.
+
+initial_max_stream_data_uni (0x000b):
+
+: The initial stream maximum data for unidirectional streams parameter contains
+  the initial flow control limit for newly created unidirectional streams opened
+  by the endpoint that receives the transport parameter.  In client transport
+  parameters, this applies to streams with an identifier ending in 0x3; in
+  server transport parameters, this applies to streams ending in 0x2.
+
+If present, transport parameters that set initial flow control limits are
+equivalent to sending a MAX_STREAM_DATA frame ({{frame-max-stream-data}}) on
+every stream of the corresponding type immediately after opening.  If the
+transport parameter is absent, streams of that type start with a flow control
+limit of 0.
```

Comment:
That's actually the only way to open a new unidirectional stream in that case. I'm not sure if I like it, since it adds another special case to my STREAM frame sending logic, since I usually never send empty STREAM frames, but I **have** to do it in this case.

Reply:
add one to the byte size of each stream packet when computing limits, or any frame contributing to flow control.

Reply:
For the HTTP/QUIC case it be more useful to have max data equivalent to 8 bits, so as to enable the creation of a stream with its type. However, I know that is not a general transport feature...

Reply:
I don't really know that *anything* would be practical here.  If you don't permit streams of this type, it's moot.  If you do permit streams of this type, you want a particular value and will probably send it as part of your settings.  But if you omit it, what is the peer allowed to send?  I don't think the goal is to pick a value that would encourage implementations to skip sending this parameter, just to not have it be undefined.

Zero is fine.  So is 4KB.  So is infinity.  Real implementations will send a value for any stream they intend to open.

Reply:
You can send STREAM_BLOCKED on any stream that you can open, but no otherwise send on.  That has the effect of opening the stream.  You don't have to send an empty STREAM frame.

As @LPardue says, there are values that might be better than others as an absolute minimum.  For instance, knowing the type of unidirectional stream is probably better than knowing nothing at all, but we have to trust that implementations provide starting flow control limits that make sense for them.  They have to deal with the consequences, after all.

0 was a valid value in the past, so we're not dealing with anything new here, except to the extent that defaults matter.